src/main/java/net/sf/marineapi/nmea/event/AbstractSentenceListener.java

```diff
@@ -101,9 +101,15 @@ public void readingStopped() {
 	@SuppressWarnings("unchecked")
 	public void sentenceRead(SentenceEvent event) {
 		Sentence sentence = event.getSentence();
-		Class<?>[] interfaces = sentence.getClass().getInterfaces();
-		if (Arrays.asList(interfaces).contains(this.expectedType)) {
-			sentenceRead((T) sentence);
+		if (this.expectedType instanceof Class<?>) {
```

Comment:
This works only for direct subclassing. The mechanism generally fails for transitive subclassing. While this may be an acceptable limitation, it should be checked and avoided, e.g., by throwing an exception.

Here is an example for making the mechanism fail:

    private static class MyGenericSentenceListener<A, B extends Sentence> extends AbstractSentenceListener<B> {
        private B lastSentence;
        
        @Override
        public void sentenceRead(B sentence) {
            lastSentence = sentence;
        }

        public B getLastSentence() {
            return lastSentence;
        }
    }
    
    private static class MyConcreteSentenceListener<A> extends MyGenericSentenceListener<A, TimeSentence> {
    }
    
    @Test
    public void testComplicatedGenericSentenceListenerSubclass() throws InterruptedException {
        MyConcreteSentenceListener<String> listener = new MyConcreteSentenceListener<String>();
        sentenceReader.addSentenceListener(listener);
        readerSupport.startReading();
        readerSupport.waitUntilAllMessagesHaveBeenRead();
        assertNotNull(listener.getLastSentence());
    }

superClass in this case will be

com.sap.sailing.nmeaconnector.test.GeneralizedSentenceListenerTest.com.sap.sailing.nmeaconnector.test.GeneralizedSentenceListenerTest$MyGenericSentenceListener<A, net.sf.marineapi.nmea.sentence.TimeSentence>

and superClassTypeArgs then is

[A, interface net.sf.marineapi.nmea.sentence.TimeSentence]

whose element [0] is a TypeVariableImpl object representing the formal type argument A of class MyConcreteSentenceListener. I guess, to make things work correctly for transitive subclasses one would need to ascend the superclass hierarchy and walking through ParameterizedType types to their raw types and from there to their superclasses until reaching the AbstractSentenceListener class. Only there can we know that the type argument [0] represents the sentence type <T>.